File: 3212-count-submatrices-with-equal-frequency-of-x-and-y/3212-count-submatrices-with-equal-frequency-of-x-and-y.java
class Solution {
    public int numberOfSubmatrices(char[][] grid) {
        int ans = 0, r = grid.length, c = grid[0].length, x = 0, y = 0, k = 0;
        int[][] arr = new int[2][c];
        for (int i = 0; i < r; i++,k=0) {
            x = 0;
            y = 0;
            for (int j = 0; j < c; j++) {
                if (grid[i][j] == 'X') {
                    x++;
                } else if (grid[i][j] == 'Y')
                    y++;
                arr[0][k] += x;
                arr[1][k] += y;
                if (arr[0][k] == arr[1][k] && arr[0][k] != 0)
                    ans++;
                k++;
            }
        }
        return ans;
    }
}

File: 3212-count-submatrices-with-equal-frequency-of-x-and-y/3212-count-submatrices-with-equal-frequency-of-x-and-y.py
class Solution:
    def numberOfSubmatrices(self, grid: List[List[str]]) -> int:
        M = len(grid)
        N = len(grid[0])

        mat = [[0] * N for _ in range(M)]

        res = 0
        for m in range(M):
            for n in range(N):
                x, y = mat[m][n - 1] if n != 0 else (0, 0)
                u, v = mat[m - 1][n] if m != 0 else (0, 0)
                a, b = mat[m - 1][n - 1] if m != 0 and n != 0 else (0, 0)

                xs, ys = (x + u - a + (grid[m][n] == 'X'), y + v - b + (grid[m][n] == 'Y'))

                mat[m][n] = (xs, ys)

                if xs == ys and xs != 0:
                    res += 1

        return res